src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,791 @@
+	}
+	return CreatePrimitiveSegment(allocator, owning_vector, capacity, input.GetType());
+}
+
+static ListSegment *GetSegment(Allocator &allocator, vector<AllocatedData> &owning_vector, LinkedList *linked_list,
+                               Vector &input) {
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new one and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+static ListSegment *GetCharSegment(Allocator &allocator, vector<AllocatedData> &owning_vector,
+                                   LinkedList *linked_list) {
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) char segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new char segment and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+static void WriteDataToSegment(Allocator &allocator, vector<AllocatedData> &owning_vector, ListSegment *segment,
+                               Vector &input, idx_t &entry_idx, idx_t &count) {
+
+	// get the vector data and the source index of the entry that we want to write
+	auto input_data = FlatVector::GetData(input);
+
+	// write null validity
+	auto null_mask = GetNullMask(segment);
+	auto is_null = FlatVector::IsNull(input, entry_idx);
+	null_mask[segment->count] = is_null;
+
+	// write value
+	if (input.GetType().InternalType() == PhysicalType::VARCHAR) {
```

Comment:
Can we make this a switch statement?
```c++
switch (input.GetType().InternalType()) {
case PhysicalType::VARCHAR:
    // stuff
case PhysicalType::LIST:
    // stuff
case PhysicalType::STRUCT:
    // physical type of logical STRUCT/MAP is both STRUCT
    // more stuff
default:
    // stuff
}
```

Also, I hate to mention this, but this seems like another good candidate for doing the type switch in the bind, rather than during execution.

This would make the code path for doing a list aggregate on primitive types much shorter